compiler/noirc_printable_type/src/lib.rs

```diff
@@ -418,14 +470,25 @@ pub fn decode_printable_value<F: AcirField>(
         PrintableType::Unit => PrintableValue::Field(F::zero()),
         PrintableType::Enum { name: _, variants } => {
             let tag = field_iterator.next().expect("not enough data: expected enum tag");
-            let tag_value = tag.to_u128() as usize;
-
-            let (_name, variant_types) = &variants[tag_value];
-            PrintableValue::Vec {
-                array_elements: vecmap(variant_types, |typ| {
-                    decode_printable_value(field_iterator, typ)
-                }),
-                is_slice: false,
+            let tag = tag.to_u128() as usize;
+            // A serialized enum looks as follows:
+            //  [tag, variant0.field0, ..., variant0.fieldN, variant1.field0, ..., variant1.fieldM, ...]
+            // So the number of fields are always the same, and we have to consume all of them;
+            // the tag tells us which ones are non-default values.
+
+            let mut elements = None;
+            for (i, (_, types)) in variants.iter().enumerate() {
+                // We have to consume all elements, so that the item that follows the enum can start at the right place in the iterator.
+                let variant_elements =
+                    vecmap(types, |typ| decode_printable_value(field_iterator, typ));
+                if i == tag {
+                    elements = Some(variant_elements);
+                }
```

Comment:
Nice! I guess one last thing we could do is to only create the vector if `i == tag`. Otherwise we can read from the iterator without building anything. Like in the rest of the comments, this is totally optional.

Reply:
Done, found a nicer way than the option as well 👍